### PR TITLE
Implemented AND&OR filters with unit tests

### PR DESCRIPTION
**Pull request by lbayliss-scottlogic**

### CODE REVIEW

File: src/main/java/com/scottlogic/filter/FilterOR.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Override
+    public List<UserPost> filter(List<UserPost> inputList) {
+        if (inputList == null) {
+            return Arrays.asList();
+        }
+        List<UserPost> outputList1 = firstFilter.filter(inputList);
+        List<UserPost> outputList2 = secondFilter.filter(inputList);
+
+        List<UserPost> outputList2copy = new ArrayList<>(outputList2);
```
Comment (KarimAllaouat) [inline src/main/java/com/scottlogic/filter/FilterOR.java]: To be consistent with the AND filter, and in the interest of readability, we could call this var orList.

File: src/main/java/com/scottlogic/filter/FilterOR.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        outputList2copy.removeAll(outputList1);
+        outputList1.addAll(outputList2copy);
+
+        return outputList1;
```
Comment (KarimAllaouat) [inline src/main/java/com/scottlogic/filter/FilterOR.java]: This is great and works. 

For me, this is slightly more difficult to read than it needs to be as we are; copying the second list, removing duplicate elements from it, then adding the remaining elements to the first list. This is how I could do it if I were say, filtering a pack of cards, but **can we take advantage of [another datatype](https://www.tutorialspoint.com/java/java_set_interface.htm) that doesn't allow duplicates?**

File: src/main/java/com/scottlogic/filter/FilterOR.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    public FilterOR(Filter filter1, Filter filter2) {
+        firstFilter = filter1;
+        secondFilter = filter2;
+    }
+
+    @Override
+    public List<UserPost> filter(List<UserPost> inputList) {
+        if (inputList == null) {
+            return Arrays.asList();
+        }
```
Comment (KarimAllaouat) [inline src/main/java/com/scottlogic/filter/FilterOR.java:25]: Do you think we should be consistent with the AND filter and have this null check there too? Or should we remove this check at the filter level and ensure that the object that calls these `filter` method never passes null?

File: src/test/java/com/scottlogic/utils/UserPostFilterData.java [hunk 4]
```diff
@@ -69,4 +70,118 @@
     public static List<UserPost> getFiltered_authorMatt_OR_keywordfunky() {
         return Arrays.asList(userPost2, userPost5, userPost6);
     }
+
+    public static List<UserPost> getFiltered_authorMatt_OR_likedPosts() {
+        return Arrays.asList(userPost2, userPost1, userPost3, userPost4, userPost6);
+    }
```
Comment (KarimAllaouat) [inline src/test/java/com/scottlogic/utils/UserPostFilterData.java]: Do these util methods get used in more than 1 test? If so I think keep them - we don't like duplication.

However, if they are used for only one test, why not just have say `List<UserPost> filteredPosts = Arrays.asList(userPost2, userPost1, userPost3, userPost4, userPost6);` in the test themselves.

Upsides would be:
- Readable. I know exactly what my expectations are for this test.
- Stops this file getting very very big when we add more and more unit tests.

What are your thoughts? Are there any obvious downsides?

File: src/test/java/com/scottlogic/utils/UserPostFilterData.java [hunk 5]
```diff
@@ -27,6 +27,7 @@
     private static UserPost userPost5 = new UserPost("This funky-name",
             OffsetDateTime.of(2001, 1, 1, 1, 1, 1, 0, ZoneOffset.UTC),
             "funky funk funk", 0);
+
     private static UserPost userPost6 = new UserPost("steve madden",
```
Comment (KarimAllaouat) [inline src/test/java/com/scottlogic/utils/UserPostFilterData.java]: Maybe this is a comment to be added to any refactor work you do:

Can these objects be final? If so, convention is to use all uppercase for naming: USER_POST_6

File: src/test/java/com/scottlogic/filter/FilterANDTest.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        FilterAND filterAND = new FilterAND(new FilterAuthor("Matt"), new FilterContentKeyword("funk"));
+        Assertions.assertEquals(Arrays.asList(), filterAND.filter(postsToFilter));
+    }
+
+    @Test
+    @DisplayName("Null input list")
+    void setFilterAND_nullInput() {
+        FilterAND filterAND = new FilterAND(new FilterAuthor("Matt"), new FilterContentKeyword("funky"));
+        Assertions.assertEquals(Arrays.asList(), filterAND.filter(null));
+    }
```
Comment (KarimAllaouat) [inline src/test/java/com/scottlogic/filter/FilterANDTest.java:203]: Nice edge cases. What happens when we construct an and filter like this?

`FilterAND filterAND = new FilterAND(null, new FilterContentKeyword("funky"));`.

Is this a relevant edge case?

**Review by KarimAllaouat — COMMENTED** [review]
Very well unit tested with some good edge cases. No major fixes at all. Added a few questions to explore.

File: src/main/java/com/scottlogic/filter/FilterOR.java (hunk 1 above)
Comment (lbayliss-scottlogic) [reply src/main/java/com/scottlogic/filter/FilterOR.java]: Implemented

File: src/main/java/com/scottlogic/filter/FilterOR.java (hunk 2 above)
Comment (lbayliss-scottlogic) [reply src/main/java/com/scottlogic/filter/FilterOR.java]: Gradually becoming aware of the importance of data types in Java! Implemented a Set() and adjusted tests so that ordering isn't an issue.

File: src/main/java/com/scottlogic/filter/FilterOR.java (hunk 3 above)
Comment (lbayliss-scottlogic) [reply src/main/java/com/scottlogic/filter/FilterOR.java:25]: My bad, have implemented the null check here too

File: src/test/java/com/scottlogic/utils/UserPostFilterData.java (hunk 4 above)
Comment (lbayliss-scottlogic) [reply src/test/java/com/scottlogic/utils/UserPostFilterData.java]: Having all these UserPosts at the top of each test class was a bit of an eyesore, and instead of settings getters for each UserPost I created methods to get the test result. I do agree with your points and I don't expect the methods to get used again, although Dillon mentioned it was more down to preference here so I went with this approach - is your approach described here used more?

File: src/test/java/com/scottlogic/utils/UserPostFilterData.java (hunk 5 above)
Comment (lbayliss-scottlogic) [reply src/test/java/com/scottlogic/utils/UserPostFilterData.java]: I did not know about this keyword - implemented

File: src/test/java/com/scottlogic/filter/FilterANDTest.java (hunk 6 above)
Comment (lbayliss-scottlogic) [reply src/test/java/com/scottlogic/filter/FilterANDTest.java:203]: I'd say it is definitely a relevant edge case - have tested on both OR/AND filters and it works appropriately!

File: src/main/java/com/scottlogic/filter/FilterOR.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        List<UserPost> outputList1 = firstFilter.filter(inputList);
+        List<UserPost> outputList2 = secondFilter.filter(inputList);
+
+        Set<UserPost> orList = new HashSet<>();
+        for (UserPost userPost : outputList1) {
+            orList.add(userPost);
+        }
+        for (UserPost userPost : outputList2) {
+            orList.add(userPost);
+        }
```
Comment (KarimAllaouat) [inline src/main/java/com/scottlogic/filter/FilterOR.java]:
```suggestion
        Set<UserPost> orList = new HashSet<>(outputList1);
        orList.addAll(outputList2);
```

File: src/main/java/com/scottlogic/filter/FilterOR.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        /*List<UserPost> outputList2copy = new ArrayList<>(outputList2);
+        outputList2copy.removeAll(outputList1);
+        outputList1.addAll(outputList2copy);*/
```
Comment (KarimAllaouat) [inline src/main/java/com/scottlogic/filter/FilterOR.java]: Unneeded comment.

File: src/test/java/com/scottlogic/filter/FilterORTest.java [hunk 9]
```diff
@@ -24,35 +24,36 @@ void filterAuthorMatt_OR_ContentKeywordfunky() {
             List<UserPost> postsToFilter = UserPostFilterData.getUnfilteredData();
             List<UserPost> expectedFilteredList = UserPostFilterData.getFiltered_authorMatt_OR_keywordfunky();
             FilterOR filterOR = new FilterOR(filterAuthor_Matt, new FilterContentKeyword("funky"));
-            Assertions.assertEquals(expectedFilteredList, filterOR.filter(postsToFilter));
+            //Assertions.assertEquals(expectedFilteredList, filterOR.filter(postsToFilter));
```
Comment (KarimAllaouat) [inline src/test/java/com/scottlogic/filter/FilterORTest.java]: Comment

File: src/test/java/com/scottlogic/filter/FilterORTest.java [hunk 10]
```diff
@@ -156,26 +157,26 @@ void filterLikedPosts_OR_likedPosts() {
...
             FilterOR filterOR = new FilterOR(filterContentKeyword, new FilterLikedPosts());
-            Assertions.assertEquals(expectedFilteredList, filterOR.filter(postsToFilter));
+            Assertions.assertTrue(expectedFilteredList.containsAll(filterOR.filter(postsToFilter)));
         }
 
         @Test
         @DisplayName("date")
         void setFilterAuthorMatt_OR_date() {
             List<UserPost> postsToFilter = UserPostFilterData.getUnfilteredData();
-            List<UserPost> expectedFitleredList = UserPostFilterData.getFilteredKeywordYou_OR_dates();
```
Comment (KarimAllaouat) [inline src/test/java/com/scottlogic/filter/FilterORTest.java]: I also missed the spelling 😅

**Review by KarimAllaouat — COMMENTED** [review]
Awesome 🚀 

As I can't seem to reply to the previous comments:
We wouldn't need getters and setters if we changed the access modifier to say public. We wouldn't need them at the top of each class either, we could import them and use them inline in the tests say:
```
@Test
@DisplayName("liked posts")
void filterBetweenDates_AND_likedPosts() {
    List<UserPost> postsToFilter = UserPostFilterData.getUnfilteredData();
    List<UserPost> expectedFitleredList = Arrays.asList(USER_POST_1, USER_POST_3);
    FilterAND filterAND = new FilterAND(filterBetweenDates, new FilterLikedPosts());
    Assertions.assertEquals(expectedFitleredList, filterAND.filter(postsToFilter));
}
```

File: src/main/java/com/scottlogic/filter/FilterOR.java (hunk 7 above)
Comment (lbayliss-scottlogic) [reply src/main/java/com/scottlogic/filter/FilterOR.java]: Need to start thinking more like this - implemented

**Review by KarimAllaouat — APPROVED** [review]
